### PR TITLE
lsp/frontend: Fix Model SourceLocation

### PR DESCRIPTION
**Pull request by jaraonthe-dot-net**
This attempts to fix various bogus hover data provided by the language server; by doing two things:
- Ignoring entities situated inside model definitions, as we cannot know if their type is only valid for one particular model invocation.
- Have proper `ExpandedLocation` data for all entities that a model invocation has been expanded to. This is the only way to tell that an entity is connected to a model invocation.

I had to fix `SourceLocation.join()` and the `MacroExpander` - I do not know if my changes cause issues elsewhere.

For testing it's best to use VSCode as lsp client, as it properly highlights the area that is being hovered on - in IntelliJ I see no such indication.

Some examples where the behavior change is apparent (all in `rv3264im.vadl`):
- line 73: `$zero45` no longer has hover
- line 231 & 511: no longer have hover that spans from model definition to model invocation
- also on line 231: `addr` still has hover showing a type that is provided by the model (CTRL-Click on it, you'll see what I mean)
- on line 20, change to `Arch64` -> lines 250-280 no longer trigger a hover ranging over the entire model invocation

### CODE REVIEW

File: vadl-frontend/main/vadl/ast/MacroExpander.java [hunk 1]
```diff
@@ -1114,7 +1117,7 @@ public Definition visit(AbiSequenceDefinition definition) {
   public Definition visit(SpecialPurposeRegisterDefinition definition) {
     return new SpecialPurposeRegisterDefinition(
         definition.purpose,
-        definition.exprs,
```
Comment (flofriday) [inline vadl-frontend/main/vadl/ast/MacroExpander.java:1131]: **praise:** Great catch.

File: vadl-frontend/main/vadl/ast/MacroExpander.java [hunk 2]
```diff
@@ -315,12 +316,14 @@ private <T extends Node> T expandNode(T node) {
...
         }
-        yield new RecordInstance(recordInstance.type, entries, recordInstance.sourceLocation);
+        yield new RecordInstance(recordInstance.type, entries,
+            copyLoc(recordInstance.sourceLocation));
       }
       case EncodingDefinition.EncsNode encs -> expandEncs(encs);
       case PlaceholderNode placeholderNode -> expand(placeholderNode);
       case MacroInstanceNode macroInstanceNode -> expand(macroInstanceNode);
       case MacroMatchNode macroMatchNode -> expand(macroMatchNode);
+      case UnOp unOp -> new UnOp(unOp.operator, copyLoc(unOp.location));
```
Comment (flofriday) [inline vadl-frontend/main/vadl/ast/MacroExpander.java:329]: **suggestion:** Let's also add BinOp here.

File: vadl-frontend/main/vadl/ast/MacroExpander.java
Comment (flofriday) [inline vadl-frontend/main/vadl/ast/MacroExpander.java:1]: **praise:** Thanks for the fixes here :)

File: vadl-lsp/main/vadl/lsp/AstFinderByPosition.java [hunk 3]
```diff
@@ -109,6 +109,23 @@ public abstract class AstFinderByPosition<N extends Node> extends RecursiveAstVi
...
+          vadl.utils.RopeList<SourceLocation.DirectLocation> expandedFrom
+        )
+      ) {
+        var outerLocation = expandedFrom.toList().getLast();
+        if (!primaryLocation.begin().isWithin(outerLocation)
+            || !primaryLocation.end().isWithin(outerLocation)) {
+          return false;
+        }
+      }
+
```
Comment (flofriday) [inline vadl-lsp/main/vadl/lsp/AstFinderByPosition.java:107]: **question:** Can you explain when the primary location would be contained within the outer location?

In my mind that should never be the case, so maybe can you provide an example here?

Because in my intuitive opinion the code should just look like:

```java
if (n.location() instanceof SourceLocation.ExpandedLocation) {
    return false;
}
```

But in general we said we're not going to ignore the definitions from macro definitions, but instead collect all the possible types and display them, like mentioned in this issue: https://github.com/OpenVADL/openvadl/issues/1016#issuecomment-4646829968

**Comment by flofriday** [comment]
Can you please explain the changes in SourceLocation?
I still don't quite get the kinds of problems the new implementation solves.
Also we should probably add tests for those cases in SourceLocationTest.

File: vadl-lsp/main/vadl/lsp/AstFinderByPosition.java (hunk 3 above)
Comment (jaraonthe-dot-net) [reply vadl-lsp/main/vadl/lsp/AstFinderByPosition.java:107]: I want to keep all considerations from #1016 out-of-scope here. But thanks for pointing out the issue, I must have missed it.

> Can you explain when the primary location would be contained within the outer location?

As far as I've seen, this applies to entities that are part of **arguments** passed to the macro invocation.
Examples:
- `rv3264im.vadl:231` -> `addr` (example from description above). This variable name refers to the `let` definition in the model definition, but as we are hovering the model invocation, the type definitely applies here (as opposed to entities located within model definitions, where this is not so straight-forward).
- The last example above (`Arch64` on line 20), then go to the `pseudo instruction` on line 274 - here various type hovers work just fine (e.g. instruction parameters). The entire thing (and all child entities) have a primary location that is situated within the expanded location.

If you prefer, of course I can simplify the filter to ignore these cases too, but I saw no reason to do that, as I'm not aware that the shown type information is incorrect or otherwise problematic.

**Comment by jaraonthe-dot-net** [comment]
> Can you please explain the changes in SourceLocation? I still don't quite get the kinds of problems the new implementation solves.

Firstly, `SourceLocation.join()` is relevant because there are AST entities that don't contain their own dedicated location field, instead compute their effective location from joining two child entities' locations.

My changes fix the behavior when two locations are joined that have a differing expansion stack. Part of the expansion stack (from the outermost model invocation inwards) may be the same, and in that case I need that part of the expansion stack to be kept in the result (to properly detect this entity is within a model definition/invocation in the first place). 
Instead, the previous implementation had the tendency to throw out the `expandedFrom` altogether, issuing a `DirectLocation` only (because it only includes the `expandedFrom` if the location stacks of both inputs are equal (or the innermost parts(!) are equal), and quite often (if not always) that is not the case).

This affects detection of locations that span from model definition to model invocation (see examples in description above): an entity joins the locations of two children together where one is part of the model definition, and the other is part of a model invocation argument, creating these weird `primaryLocation`s, but by dropping the expansion stack, it is very difficult to figure this out (and effectively ignore them).

Further example: `rv3264im.vadl:495`:
`let MinInt = ((1 as Regs) << ((match : Id ($opTy = SIntW => WLen; _  => MLen)) - 1)) as $opTy in`
The `match` statement creates one more entry in the expansion stack (for the entry that is the result of this `match`), which is then joined with the following `1` literal to make up the location of the subtraction operation (i.e. location with 3 `expandedFrom` locations joined with another that has only 2 (see screenshot)). The resulting mangled location has a weird `primaryLocation` and no `expandedFrom` (spanning yet again from within the model definition to within the model invocation).

If you need more convincing, we should probably discuss this in a meeting.

**EDIT**:
Two Screenshots illustrating the last example - this is without my changes to `SourceLocation`. `BinExpr` joins its left and right operand locations together, the resulting location (at the very bottom) does not contain any `expandedFrom` anymore, thus not being recognized as being part of a model invocation, leading to the bogus hover.
Without fixing SourceLocation Screenshot_20260811_225830
Without fixing SourceLocation Screenshot_20260811_230353

**Comment by jaraonthe-dot-net** [comment]
@flofriday I've added your suggestions (see Fixup commit). Please take another look.

**Comment by flofriday** [comment]
Now that I've a better tool to visualize/debug sourcelocations was the intent of this PR to fix cases like this (from the old version where it clearly is broken):

Screenshot 2026-08-19 at 20 24 58

Any other cases this intends to fix?

**Comment by jaraonthe-dot-net** [comment]
> [...] was the intent of this PR to fix cases like this [...]

Without looking at the SourceLocations involved I am not sure why this example behaves like that (either the `expandedFrom` stacks diverge -> `join()`; or some location is not properly set by `MacroExpander`). But yeah, locations like that are what I want to filter out.

> Any other cases this intends to fix?

I'm trying to fix these issues:
- SourceLocation spanning from model definition to model invocation (some of those cannot be helped (i.e. an entity that joins two children where one is part of the definition and the other an invocation argument), thus the explicit filtering of these). (Your example falls into this category)
- SourceLocations that should have an `expandedFrom` stack but don't (because `join()` drops it). This is a very big problem because it makes it impossible to detect that a model is being used in the first place.
- The outermost `expandedFrom` becoming the effective location (my last example in the description).

File: vadl-lsp/main/vadl/lsp/AstFinderByPosition.java (hunk 3 above)
Comment (flofriday) [reply vadl-lsp/main/vadl/lsp/AstFinderByPosition.java:107]: Yeah that makes sense, nice catch.

File: vadl/main/vadl/utils/SourceLocation.java [hunk 4]
```diff
@@ -161,47 +161,42 @@ default SourceLocation join(SourceLocation other) {
...
+
+        return SourceLocation.of(firstLocation.path(), begin, end, expandedFrom);
       }
     }
 
-    if (!Objects.equals(firstLocation.path(), secondLocation.path())) {
+    // All minSize outer layers are the same for both locations; let's return them
+    if (!Objects.equals(thisStack.get(minSize - 1).path(), otherStack.get(minSize - 1).path())) {
       throw new IllegalArgumentException("Cannot join source locations from different files.");
     }
```
Comment (flofriday) [inline vadl/main/vadl/utils/SourceLocation.java]: **question:** Is this condition necessary, doesn't the loop already check if for all i < minSize?

**Comment by flofriday** [comment]
I've went over the fix again now (with the Advanced AST Dump merged in #1067) and your implementation seems to be quite correct. I cannot come up with any counter-example and all the edge cases I tested are now correct or at least not more broken than before, well done ✨

However, I want to have more and easier to read tests for example this code is a minimal example of a previously really wrongly reported error message. That would make a nice addition to frontend  snapshottests (probably in the typechecker).
```
// Before #1052 this code reported a wrong source location.
model abc() : Ex = { 0 }

model xyz(): Defs = {
  constant flo = (5 + $abc) / $abc
}

$xyz()
```

On master this reports as:
```
error: Constant value required
     ╭── file:///Users/flo/src/open-vadl/example.vadl:5:19
     │
   5 │>   constant flo = (5 + $abc) / $abc
   6 │> }
   7 │> 
   8 │> $xyz()
     │    Division by zero cannot be computed.
     │
```

which is not correct and also dropped the expanded-from stack.
But with your change it reports only the division as expected 🎉


**Update:** While playing around I discovered another issue in the model expander which I wrote down here: #1075
If you want you can also try to fix it in this PR or in another one.

**Comment by jaraonthe-dot-net** [comment]
@flofriday I've added the snapshot test you wanted me to add. Anything else before we can finish this PR?

**Comment by flofriday** [comment]
Yes, there is still an unresolved conversation about a condition.

**Comment by jaraonthe-dot-net** [comment]
> Yes, there is still an unresolved conversation about a condition.

I don't see any unresolved conversations on this PR. What are you referring to?

**Comment by flofriday** [comment]
You're right I forgot to publish them, my bad sorry.

File: vadl/main/vadl/utils/SourceLocation.java (hunk 4 above)
Comment (jaraonthe-dot-net) [reply vadl/main/vadl/utils/SourceLocation.java]: You're right, it's not necessary. Removed now.

**Review by flofriday — APPROVED** [review]
LGTM
